akka-remote/src/main/scala/akka/remote/artery/compress/TopHeavyHitters.scala

```diff
@@ -56,30 +72,48 @@ private[remote] final class TopHeavyHitters[T](val max: Int) {
    */
   // TODO possibly can be optimised further? (there is a benchmark)
   def update(item: T, count: Long): Boolean =
-    isHeavy(count) && { // O(1) terminate execution ASAP if known to not be a heavy hitter anyway 
+    isHeavy(count) && { // O(1) terminate execution ASAP if known to not be a heavy hitter anyway
       val hashCode = new HashCodeVal(item.hashCode()) // avoid re-calculating hashCode
-      (findHashIdx(0, hashCode): @switch) match { // worst case O(n), can't really bin search here since indexes are kept in synch with other arrays hmm...  
+      val startIndex = hashCode.get & mask
+      (findHashIdx(startIndex, hashCode): @switch) match { // worst case O(n), common O(1 + alpha), can't really bin search here since indexes are kept in synch with other arrays hmm...
         case -1 ⇒
           // not previously heavy hitter
-          insertKnownNewHeavy(hashCode, item, count) // O(1) + rarely O(n) if needs to update lowest hitter
-
+          insertKnownNewHeavy(hashCode, item, count) // O(log n + alpha)
```

Comment:
And still pessimistically could be O(n), but in common case where we have now \alpha = 0.5 is O(1)